crates/biome_js_type_info/src/flattening/intersections.rs

```diff
@@ -0,0 +1,319 @@
+use biome_rowan::Text;
+
+use crate::{
+    Class, Function, Interface, Intersection, Literal, Namespace, Object, ResolvedTypeData,
+    ReturnType, TypeData, TypeMember, TypeResolver,
+};
+
+pub(super) fn flattened_intersection(
+    intersection: &Intersection,
+    resolver: &mut dyn TypeResolver,
+) -> TypeData {
+    match intersection.types() {
+        [] => TypeData::NeverKeyword,
+        [ty] => resolver
+            .resolve_and_get(ty)
+            .map_or_else(TypeData::unknown, ResolvedTypeData::to_data),
+        types => {
+            let merged_ty = resolver
+                .resolve_and_get(&types[0])
+                .map_or_else(TypeData::unknown, ResolvedTypeData::to_data);
+            let mut merged_ty = MergedType::from_type(merged_ty, resolver);
+            let mut primitive = merged_ty.as_primitive().cloned();
+            for other in &types[1..] {
+                if !merged_ty.is_mergeable() {
+                    break;
+                }
+
+                let other_ty = resolver
+                    .resolve_and_get(other)
+                    .map_or_else(TypeData::unknown, ResolvedTypeData::to_data);
+                let other_ty = MergedType::from_type(other_ty, resolver);
+                if let Some(other_primitve) = other_ty.as_primitive() {
+                    if primitive.is_some() {
+                        return TypeData::NeverKeyword;
+                    }
+
+                    primitive = Some(other_primitve.clone());
+                } else {
+                    merged_ty = merged_ty.intersection_with(other_ty, resolver);
+                }
+            }
+
+            if let Some(primitive) = primitive {
+                primitive
+            } else {
+                merged_ty.into_type()
+            }
+        }
+    }
+}
+
+enum MergedType {
```

Comment:
Should we add arrays? Geniunely asking!

Reply:
Array's are stored as instances of the `Array` class, so technically they could already be covered by this, although in practice we don't and we may indeed need special handling for them.

Tuples do have their own variant that we also may have want to support at some point. I'll add a comment for them!